toc/rfc/rfc-draft-cnb-lifecycle.md

```diff
@@ -0,0 +1,135 @@
+  - No OCI registry necessary
+  - Reuse cflinuxfs4 as rootfs during build and run
+
+### High Level Implementation
+
+#### CNB App Lifecycle
+
+Introduce a new `lifecycle type` that enables the cloud controller to differentiate between the classical buildpacks and Cloud Native Buildpacks. On a high level, it will be very similar to the existing buildpackapplifecycle (v2 buildpacks). The new app lifecycle acts as a CNB [platform](https://buildpacks.io/docs/for-app-developers/concepts/platform/) and will:
+
+1. Download the app source code from the blobstore
+1. Download the CNB app lifecycle from the blobstore
+1. Download the configured buildpacks
+1. Write an [order.toml](https://github.com/buildpacks/spec/blob/main/platform.md#ordertoml-toml) with configured buildpacks
+1. Execute [detect](https://github.com/buildpacks/spec/blob/main/platform.md#detector) and [build](https://github.com/buildpacks/spec/blob/main/platform.md#builder) phases using the [CNB lifecycle](https://github.com/buildpacks/lifecycle)
+1. Package the result into a droplet and upload it to the blobstore
+1. Write a result.json file with the [Staging Result](https://github.com/cloudfoundry/buildpackapplifecycle/blob/f4b2bc9ff6cc6229402d7c27e887763154cf0378/models.go#L73-L80)
+
+#### CNB Lifecycle Type
+
+Introduce a new type of lifecycle type which indicates that Cloud Native Buildpacks should be used. In future, this can be enhanced with additional CNB inputs. For this RFC we’d start with:
+
+```json
+{
+  "type": "cnb",
+  "data": {
+    "buildpacks": ["gcr.io/paketo-buildpacks/java"],
+    "stack": "cflinuxfs4"
+  }
+}
+```
+
+Both, building and running an app will be based on the configured stack. If no stack is provided, the platform default is used. An empty (or not provided) list of buildpacks will lead to an error. This essentially means, that no auto-detection is supported at the moment. Once [system CNBs](#system-buildpacks) are supported, this behavior will change.
+
+#### CF CLI
+
+New flag `–[no-]cnb` will be introduced to the `cf push` command. The flag can not be combined with any ` --docker-*` commands. When the flag is set, CF CLI must set the lifecycle type to `cnb`.
+
+#### App Manifest
+
+New property `cnb: true|false` will be added to the App manifest.
+
+```yaml
+---
+applications:
+  - name: test-app
+    instances: 1
+    cnb: true
+    buildpacks:
+      - gcr.io/paketo-buildpacks/java
+```
+
+The `cnb` property cannot be present together with any of the `docker-*` properties. When the flag is set to `true`, the CF CLI must set the lifecycle type to `cnb`.
+
+Both changes (CLI and manifest) were chosen because they are simple (from a user perspective), easy to implement and remove, if CNBs will become the standard lifecycle in future.
+
+### Alternative APIs
+
+- Instead of a binary switch, introduce a `buildpack-type` (`v2`/`v3` or `cf`/`cnb` or `classic`/`cnb`) to distinguish between different lifecycles
```

Comment:
+1 for `buildpack-type` in app manifest. It allows smooth extension in the future while a `cnb: true` means that every new lifecycle type needs additional properties. Future extensions might be a CNB lifecycle that builds OCI images (including stack) instead of traditional droplets (as @FloThinksPi suggested below).

`buildpack-type` matches to a certain extent the `lifecycle.type` attribute in the [CF API](https://v3-apidocs.cloudfoundry.org/index.html#lifecycles). Full consistency between CF API, app manifest and cf cli is probably out of reach due to history.

Reply:
As you explicitly mention a potential CNB based lifecycle that introduces full OCI images, we should then probably not use `cnb`, but rather something like `cnb-droplet`. WDYT?

Reply:
I would also prefer to have `buildpack-type` instead of cnb: true, which more or less looks like a switch.

Reply:
We've updated the RFC and changed the `cnb: true|false` to the `lifecycle: buildpack|docker|cnb`. However, the introduction of `cnb-oci` or `cnb-droplet` seems to be too early. What would be the benefit for the end user to be able to choose between droplet-style and raw OCI? But since the whole feature will be behind a feature flag, the behaviour could be also controller by administrators on the CCNG config level.

Reply:
Potential end user benefit:
- cnb-droplet: user gets stack updates automatically, similar to buildpack lifecycle
- cnb-oci: user gets a standard OCI image that could be used outside outside of CF as well (ok, a rather artificial benefit as there are other options to build such an image), but no automatic stack updates (need a restage) 

So yes, one cnb lifecycle should be sufficient for today.